MAINT: gracefully shuffle memoryviews

tylerjereddy opened the pull request:
Fixes gh-18273

* allow graceful shuffling of memoryviews, with
same behavior as arrays, instead of producing
a warning on `memoryview` shuffle

There's certainly some contention that me doing this downstream was sub-optimal anyway, but I don't think there's any claim that it should be disallowed/considered a bug to shuffle a memoryview.

I was originally going to special case 1-D C-contiguous memoryviews, but the discussion in the matching issue seems to suggest that just using `asarray()` without fancy special casing might be the way to go?

tylerjereddy committed:
MAINT: gracefully shuffle memoryviews

* allow graceful shuffling of memoryviews, with
same behavior as arrays, instead of producing
a warning on `memoryview` shuffle

tylerjereddy commented on numpy/random/_generator.pyx:
```diff
@@ -4398,6 +4398,9 @@ cdef class Generator:
             char* x_ptr
             char* buf_ptr
 
+        if isinstance(x, memoryview):
```
I'm guessing this conditional could just be removed? --was being conservative I suppose

rkern replied:
Since `x` can be a `list` that we definitely do not want to view as an `ndarray`, this conditional is not optional.

eric-wieser commented:
I wonder if we should instead change the `MutableSequence` check to `Sequence`? That would allow that path to automatically work for `memoryview`.

rkern commented:
> There's certainly some contention that me doing this downstream was sub-optimal anyway, but I don't think there's any claim that it should be disallowed/considered a bug to shuffle a memoryview.

That was specific to that particular use of memoryviews post-shuffle, not the shuffling per se. This is a good change.

seberg commented:
Both changes seem fine to me (`isinstance` check or Sequence).  Since a non-mutable sequence should error already, I somewhat doubt that modifying that as Eric suggested has any downsides. Presumably, any of the tensor frameworks will either be both or neither.  So I think Eric's suggestion is probably good.

I am not quite sure that the memoryview change can't go wrong in principle (memoryviews do support suboffsets for example), but I am willing to take the risk of such an unlikely backcompat break (it would go to an error), it probably is nicer for memoryviews. Although, I doubt there is a reason to optimize for them here.

tylerjereddy committed:
MAINT: PR 18282 revisions

* based on reviewer feedback, instead of explicitly
supporting `memoryview` shuffling via `asarray()`, support
the shuffling implicitly by using `Sequence` instead
of `MutableSequence`

tylerjereddy commented:
I pushed in Eric's suggestion after confirming the random suite still passes locally for this feature branch.

seberg commented:
Thanks, LGTM. I find it a bit surprising that a memoryview isn't considered a mutable sequence (is that supposed to be only the length?), but I guess if Python changes it, we have a test that notifies us.

I would be fine with a memoryview fast path as you had it, although I guess this might be nicer for backporting.

eric-wieser commented:
`MutableSequence` has to implement `insert`, which `memoryview` would never do.

seberg commented:
Ah right. So I think this is still good enough to warn for tensorflow as well? In which case I think we should go with it or is there some other reason to prefer MutableSequence, that I am missing?

Although, I would be happy to entertain the thought of making things more strict in 1.21.

charris commented:
Thanks Tyler.
